### ISSUE TITLE
Exhaustive if else if check on union types is still asking for return statement

### ISSUE THREAD

**Issue by safareli** [microsoft/TypeScript#38731]
**TypeScript Version:**  3.9.2

**Search Terms:** exhaustive if else if is resulting in Function lacks ending return statement and return type does not include 'undefined'.(2366)

**Code**

```ts
export const getPos = (ab: "A" | "B"): 1 => {
    if (ab == "A") {
        return 1;
    } else if (ab == "B") {
        return 1;
    }
    // if you hover on `ab`, type is `never`. 
    const proveIsNever: never = ab;
    // if you uncomment next line, error will disappear.
    // return proveIsNever;
}
```

**Expected behavior:** is to get no error, as type of `ab` after the if/else if is `never` (or it's type shouldn't be `never`).

**Actual behavior:** is an error: `Function lacks ending return statement and return type does not include 'undefined'.(2366)`

**Playground Link:** [Provided](https://www.typescriptlang.org/play/index.html?ssl=1&ssc=1&pln=12&pc=1#code/KYDwDg9gTgLgBAYwgOwM7wObBgBQquAXjgAoBDAIwC44AiAQVrgB86AhWgShoEYiA+OAG8AUHHFwAlgDNSlIsQZdhYiWqjYArlGRweAblXiAvnGAAbVMCmzyFBe2Wi16rTr2G1xo3AD0vmzgATwhNOAALCAA3YCg4FDgAA0pEgBo4GCCwa0kCRORgGKhEgDo4HyQ0eDAoaOAASVQAOULYmgKiojhKTwl-QJCwzWQkAFtR4GR4ApB4c0kC9NjauIB3SXNzOAATXLIwbLIoEp9+jRhtXRq6xpaiw28gA)

**Related Issues:** #36884 and #28520 are related but in those case one might argue that you can't have exhaustive check using instanceOf due to structural typing, but in this case it's definitely possible. or if it's not possible `ab` shouldn't have type `never` in the end.

**Comment by safareli** [comment]
This work as expected tho:
```

export const getPos = (ab: "A" | "B"): 1 => {
    if (ab == "A") {
        return 1;
    } else {
        return 1;
    }
}



export const getPos_ = (ab: "A" | "B"): 1 => {
    switch (ab) {
        case "A":
            return 1;
        case "B":
            return 1;
    }
}
```

**Comment by ahejlsberg** [comment]
Duplicate of #21985.
